src/Servers/Kestrel/perf/Kestrel.Performance/SchedulerBenchmark.cs

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Concurrent;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using BenchmarkDotNet.Attributes;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Performance
+{
+    public class SchedulerBenchmark
+    {
+        private const int InnerLoopCount = 1024;
+        private const int OuterLoopCount = 64;
+        private const int OperationsPerInvoke = InnerLoopCount * OuterLoopCount;
+
+        private readonly int IOQueueCount = Math.Min(Environment.ProcessorCount, 16);
+
+        private PipeScheduler[] _lockFreeQueue;
+        private PipeScheduler[] _lockBasedQueue;
+        private PipeScheduler _threadPoolScheduler;
+
+        private static Action<object> _action = (o) => { };
+
+        private static Action<int> _lockFreeAction;
+        private static Action<int> _lockBasedAction;
+        private static Action<int> _threadPoolAction;
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            _lockFreeQueue = new IOQueueLockFree[IOQueueCount];
+            for (var i = 0; i < _lockFreeQueue.Length; i++)
+            {
+                _lockFreeQueue[i] = new IOQueueLockFree();
+            }
+
+            _lockFreeAction =
+                (n) =>
+                {
+                    PipeScheduler pipeScheduler = _lockFreeQueue[n % _lockFreeQueue.Length];
+                    for (var i = 0; i < InnerLoopCount; i++)
+                    {
+                        pipeScheduler.Schedule(_action, null);
+                    }
+                };
```

Comment:
It would read a little cleaner to define these actions as methods.

Reply:
Done